File: lib/screenshotParser.js
import {
  parseDistanceM,
  parseDurationSec,
  parsePaceSecPer100m,
} from './swimFormatters.js';

const DUTCH_DAYS = {
  ma: 1, di: 2, wo: 3, do: 4, vr: 5, za: 6, zo: 0,
};

const DUTCH_MONTHS = {
  jan: 0, feb: 1, mrt: 2, mar: 2, apr: 3, mei: 4, may: 4,
  jun: 5, jul: 6, aug: 7, sep: 8, okt: 9, oct: 9, nov: 10, dec: 11,
};

const normalizeText = (text) => String(text || '')
  .replace(/\u2013|\u2014/g, '-')
  .replace(/[''′`]/g, "'")
  .replace(/"/g, '"')
  .replace(/\s+/g, ' ')
  .trim();

const firstMatch = (text, patterns) => {
  for (const pattern of patterns) {
    const match = text.match(pattern);
    if (match) return match;
  }
  return null;
};

export const parseDutchDate = (text, referenceDate = new Date()) => {
  const match = text.match(
    /\b(ma|di|wo|do|vr|za|zo)\s+(\d{1,2})\s+(jan|feb|mrt|mar|apr|mei|may|jun|jul|aug|sep|okt|oct|nov|dec)\b/i
  );
  if (!match) return null;

  const day = parseInt(match[2], 10);
  const monthKey = match[3].toLowerCase().slice(0, 3);
  const month = DUTCH_MONTHS[monthKey];
  if (month == null || day < 1 || day > 31) return null;

  let year = referenceDate.getFullYear();
  const candidate = new Date(year, month, day);
  if (candidate > referenceDate) {
    year -= 1;
  }
  const iso = `${year}-${String(month + 1).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
  return iso;
};

const parseStrokeDistance = (text, labelPattern) => {
  const match = text.match(new RegExp(`${labelPattern}\\s*\\(([\\d.\\s]+)\\s*m\\)`, 'i'));
  if (!match) return null;
  return parseDistanceM(match[1]);
};

const parseOcrPace = (raw) => {
  if (!raw) return null;
  const direct = parsePaceSecPer100m(raw);
  if (direct) return direct;
  const compact = String(raw).match(/^(\d{2,3})"/);
  if (compact) {
    const digits = compact[1];
    if (digits.length === 3) {
      return parseInt(digits[0], 10) * 60 + parseInt(digits.slice(1), 10);
    }
  }
  return null;
};

const parseDistanceToken = (token) => {
  if (!token) return null;
  return parseDistanceM(token.replace(/\s/g, ''));
};

/** Apple Fitness activity labels and metrics that indicate a non-swim workout. */
const OTHER_SPORT_PATTERNS = [
  { sport: 'run', patterns: [/\bHardlopen\b/i, /\bIndoor\s*Run\b/i, /\bJoggen\b/i, /\bLoopband\b/i] },
  { sport: 'walk', patterns: [/\bWandelen\b/i, /\bOutdoor\s*Walk\b/i, /\bWandeltocht\b/i] },
  { sport: 'cycle', patterns: [/\bFietsen\b/i, /\bIndoor\s*Cycling\b/i, /\bOutdoor\s*Cycling\b/i] },
  { sport: 'row', patterns: [/\bRoeien\b/i, /\bIndoor\s*Rowing\b/i] },
  { sport: 'hike', patterns: [/\bHiking\b/i, /\bBergwandelen\b/i] },
  { sport: 'yoga', patterns: [/\bYoga\b/i, /\bPilates\b/i] },
  { sport: 'strength', patterns: [/\bKrachttraining\b/i, /\bFunctional\s*Strength\b/i, /\bTraditionele\s*krachttraining\b/i] },
  { sport: 'elliptical', patterns: [/\bCrosstrainer\b/i, /\bElliptical\b/i] },
  { sport: 'stair', patterns: [/\bTraplopen\b/i, /\bStair\s*Stepper\b/i] },
  { sport: 'dance', patterns: [/\bDans\b/i, /\bDance\b/i] },
];

const NON_SWIM_METRIC_PATTERNS = [
  /\bStappen\b/i,
  /\bHoogtemeters\b/i,
  /\bElevation\b/i,
  /\bAfstand\s*[\d.,]+\s*KM\b/i,
  /\/\s*KM\b/i,
  /\bmin\s*\/\s*km\b/i,
  /\bRPM\b/i,
  /\bWATT\b/i,
  /\bCadans\b/i,
  /\bBPM\b/i,
];

export const detectSwimWorkout = (text, fields = {}) => {
  const normalized = normalizeText(String(text || '').replace(/\n/g, ' '));

  for (const { sport, patterns } of OTHER_SPORT_PATTERNS) {
    if (patterns.some((p) => p.test(normalized))) {
      return { isSwimWorkout: false, detectedSport: sport };
    }
  }

  const swimSignals = [
    /\bZwem(bad|men)\b/i.test(normalized),
    /\bOpen\s*water\s*zwemmen\b/i.test(normalized),
    /\bBanen\b/i.test(normalized),
    /\bBaanlengte\b/i.test(normalized),
    /\/\s*100\s*m\b/i.test(normalized),
    fields.paceSecPer100m != null,
    /\b(Schoolslag|Vrije\s*slag|Rugslag|Vlinderslag)\b/i.test(normalized),
    fields.laps != null,
    /Doel:\s*[\d.]+\s*M\b/i.test(normalized),
    Object.values(fields.strokes || {}).some((v) => v != null),
  ].filter(Boolean).length;

  const hasNonSwimMetrics = NON_SWIM_METRIC_PATTERNS.some((p) => p.test(normalized));
  if (hasNonSwimMetrics && swimSignals < 2) {
    return { isSwimWorkout: false, detectedSport: 'other' };
  }

  if (swimSignals >= 2) {
    return { isSwimWorkout: true, detectedSport: null };
  }

  return { isSwimWorkout: false, detectedSport: 'unknown' };
};

export const parseScreenshotText = (ocrText, referenceDate = new Date()) => {
  const raw = String(ocrText || '');
  const text = normalizeText(raw.replace(/\n/g, ' '));
  const warnings = [];
  let confidence = 0;
  let fieldsFound = 0;

  const date = parseDutchDate(text, referenceDate);
  const missingDate = !date;
  if (date) fieldsFound += 1;

  const durationMatch = firstMatch(text, [
    /Work[- ]?outtijd\s*Afstand\s*(\d+:\d{2}:\d{2})/i,
    /Work[- ]?outtijd\s*(\d+:\d{2}:\d{2})/i,
    /\b(\d+:\d{2}:\d{2})\b/,
  ]);
  const durationSec = durationMatch ? parseDurationSec(durationMatch[1]) : null;
  if (durationSec) fieldsFound += 1;

  const distanceMatch = firstMatch(text, [
    /\d+:\d{2}:\d{2}\s+([\d.]+\s*M)/i,
    /Afstand\s*([\d.]+\s*M)/i,
    /Distance\s*([\d.]+\s*M)/i,
  ]);
  const distanceM = distanceMatch ? parseDistanceToken(distanceMatch[1]) : null;
  if (distanceM) fieldsFound += 1;

  let activeKcal = null;
  let totalKcal = null;
  const dualKcalMatch = text.match(
    /Actieve\s*kilocalorie[ëe]n\s*Totale\s*kilocalorie[ëe]n\s*(\d+)\s*KCAL\s*(\d+)\s*KCAL/i
  );
  if (dualKcalMatch) {
    activeKcal = parseInt(dualKcalMatch[1], 10);
    totalKcal = parseInt(dualKcalMatch[2], 10);
  } else {
    const activeKcalMatch = firstMatch(text, [
      /Actieve\s*kilocalorie[ëe]n\s*(\d+)/i,
      /Active\s*calories\s*(\d+)/i,
    ]);
    activeKcal = activeKcalMatch ? parseInt(activeKcalMatch[1], 10) : null;
    const totalKcalMatch = firstMatch(text, [
      /Totale\s*kilocalorie[ëe]n\s*(\d+)/i,
      /Total\s*calories\s*(\d+)/i,
    ]);
    totalKcal = totalKcalMatch ? parseInt(totalKcalMatch[1], 10) : null;
  }
  if (activeKcal) fieldsFound += 1;
  if (totalKcal) fieldsFound += 1;

  const paceMatch = firstMatch(text, [
    /Gem\.?\s*tempo\s*Gem\.?\s*hartslag\s*(\d+'\d{1,2}")/i,
    /Gem\.?\s*tempo\s*(\d+'\d{1,2}")/i,
    /Gem\.?\s*tempo\s*Gem\.?\s*hartslag\s*(\d{2,3}")/i,
    /(\d+'\d{1,2}")(?:\s*\/\s*100\s*m|\s*\d*\s*m)/i,
    /(\d{2,3}")(?:\s*[n\/]?\d*\s*m|\s*SPM)/i,
  ]);
  const paceSecPer100m = paceMatch ? parseOcrPace(paceMatch[1]) : null;
  if (paceSecPer100m) fieldsFound += 1;

  const hrMatch = firstMatch(text, [
    /(\d{2,3})\s*SPM/i,
    /Gem\.?\s*hartslag\s*(\d{2,3})/i,
  ]);
  const avgHeartRate = hrMatch ? parseInt(hrMatch[1], 10) : null;
  if (avgHeartRate) fieldsFound += 1;

  let laps = null;
  let poolLengthM = 25;
  const lapsPoolMatch = text.match(/Banen\s*Baanlengte\s*(\d+)\s*(\d+)\s*M/i);
  if (lapsPoolMatch) {
    laps = parseInt(lapsPoolMatch[1], 10);
    poolLengthM = parseInt(lapsPoolMatch[2], 10);
  } else {
    const lapsMatch = firstMatch(text, [/Banen\s*(\d+)/i, /Laps\s*(\d+)/i]);
    laps = lapsMatch ? parseInt(lapsMatch[1], 10) : null;
    const poolMatch = firstMatch(text, [
      /Baanlengte\s*(\d+)\s*M/i,
      /\b(\d{2})\s*M\s*$/i,
    ]);
    if (poolMatch) poolLengthM = parseInt(poolMatch[1], 10);
  }
  if (laps) fieldsFound += 1;
  if (poolLengthM !== 25 || lapsPoolMatch) fieldsFound += 1;

  const goalMatch = text.match(/Doel:\s*([\d.]+)\s*M/i);
  const goalM = goalMatch ? parseDistanceM(goalMatch[1]) : null;
  if (goalM) fieldsFound += 1;

  const timeRangeMatch = text.match(/\b(\d{1,2}:\d{2})\s*-\s*(\d{1,2}:\d{2})\b/);
  const timeRange = timeRangeMatch ? `${timeRangeMatch[1]}–${timeRangeMatch[2]}` : '';

  const locationMatch = text.match(/(?:📍|pin|locatie)?\s*([A-Z][a-zà-ü]+(?:\s+[A-Z][a-zà-ü]+)?)\s*(?:\n|$|\d{1,2}:\d{2})/);
  let location = '';
  if (text.includes('Tilburg')) location = 'Tilburg';
  else if (locationMatch && !['Zwembad', 'Gemengd', 'Schoolslag'].includes(locationMatch[1])) {
    location = locationMatch[1];
  }

  const strokes = {
    mixedM: parseStrokeDistance(text, 'Gemengd'),
    breaststrokeM: parseStrokeDistance(text, 'Schoolslag'),
    freestyleM: parseStrokeDistance(text, 'Vrije\\s*slag'),
    backstrokeM: parseStrokeDistance(text, 'Rugslag'),
    butterflyM: parseStrokeDistance(text, 'Vlinderslag'),
  };

  if (Object.values(strokes).some((v) => v != null)) fieldsFound += 1;

  confidence = Math.min(100, Math.round((fieldsFound / 10) * 100));

  if (!distanceM) warnings.push('distance_not_found');
  if (!paceSecPer100m) warnings.push('pace_not_found');
  if (missingDate) warnings.push('date_missing');

  const fields = {
    date,
    durationSec,
    distanceM,
    activeKcal,
    totalKcal,
    paceSecPer100m,
    avgHeartRate,
    laps,
    poolLengthM,
    goalM,
    location,
    timeRange,
    strokes,
  };

  const { isSwimWorkout, detectedSport } = detectSwimWorkout(text, fields);
  if (!isSwimWorkout) warnings.push('not_swim_workout');

  return {
    fields,
    confidence,
    missingDate,
    warnings,
    isSwimWorkout,
    detectedSport,
  };
};

export const fieldsToSessionMetrics = (fields) => ({
  durationSec: fields.durationSec,
  distanceM: fields.distanceM,
  activeKcal: fields.activeKcal,
  totalKcal: fields.totalKcal,
  paceSecPer100m: fields.paceSecPer100m,
  avgHeartRate: fields.avgHeartRate,
  laps: fields.laps,
  poolLengthM: fields.poolLengthM || 25,
  goalM: fields.goalM,
  location: fields.location || '',
  timeRange: fields.timeRange || '',
  strokes: {
    mixedM: fields.strokes?.mixedM ?? null,
    breaststrokeM: fields.strokes?.breaststrokeM ?? null,
    freestyleM: fields.strokes?.freestyleM ?? null,
    backstrokeM: fields.strokes?.backstrokeM ?? null,
    butterflyM: fields.strokes?.butterflyM ?? null,
  },
});

/** @internal */
export const __testing = { parseDutchDate, normalizeText, detectSwimWorkout };

File: ios/AapSC/Lib/ScreenshotParser.swift
import Foundation

enum ScreenshotParser {
    private static let dutchMonths: [String: Int] = [
        "jan": 1, "feb": 2, "mrt": 3, "mar": 3, "apr": 4, "mei": 5, "may": 5,
        "jun": 6, "jul": 7, "aug": 8, "sep": 9, "okt": 10, "oct": 10, "nov": 11, "dec": 12
    ]

    static func normalizeText(_ text: String) -> String {
        text
            .replacingOccurrences(of: "\u{2013}", with: "-")
            .replacingOccurrences(of: "\u{2014}", with: "-")
            .replacingOccurrences(of: "\u{2019}", with: "'")
            .replacingOccurrences(of: "\u{2032}", with: "'")
            .replacingOccurrences(of: "`", with: "'")
            .replacingOccurrences(of: "\u{201C}", with: "\"")
            .replacingOccurrences(of: "\u{201D}", with: "\"")
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func parseDutchDate(_ text: String, referenceDate: Date = Date()) -> String? {
        let pattern = #"\b(ma|di|wo|do|vr|za|zo)\s+(\d{1,2})\s+(jan|feb|mrt|mar|apr|mei|may|jun|jul|aug|sep|okt|oct|nov|dec)\b"#
        guard let match = text.range(of: pattern, options: [.regularExpression, .caseInsensitive]) else {
            return nil
        }
        let token = String(text[match])
        let parts = token.split(separator: " ")
        guard parts.count >= 3,
              let day = Int(parts[1]),
              let month = dutchMonths[String(String(parts[2]).lowercased().prefix(3))] else {
            return nil
        }

        let calendar = Calendar.current
        var year = calendar.component(.year, from: referenceDate)
        var components = DateComponents(year: year, month: month, day: day)
        guard var candidate = calendar.date(from: components) else { return nil }
        if candidate > referenceDate {
            year -= 1
            components.year = year
            candidate = calendar.date(from: components) ?? candidate
        }
        return String(format: "%04d-%02d-%02d", year, month, day)
    }

    static func parseScreenshotText(_ ocrText: String, referenceDate: Date = Date()) -> ParsedScreenshotResult {
        let text = normalizeText(ocrText.replacingOccurrences(of: "\n", with: " "))
        var fieldsFound = 0
        var warnings: [String] = []

        let date = parseDutchDate(text, referenceDate: referenceDate)
        let missingDate = date == nil
        if date != nil { fieldsFound += 1 }

        let durationSec = firstMatch(text, patterns: [
            #"Work[- ]?outtijd\s*Afstand\s*(\d+:\d{2}:\d{2})"#,
            #"Work[- ]?outtijd\s*(\d+:\d{2}:\d{2})"#,
            #"\b(\d+:\d{2}:\d{2})\b"#
        ]).flatMap { SwimFormatters.parseDurationSec($0) }
        if durationSec != nil { fieldsFound += 1 }

        let distanceM = firstMatch(text, patterns: [
            #"\d+:\d{2}:\d{2}\s+([\d.]+\s*M)"#,
            #"Afstand\s*([\d.]+\s*M)"#,
            #"Distance\s*([\d.]+\s*M)"#
        ]).flatMap { parseDistanceToken($0) }
        if distanceM != nil { fieldsFound += 1 }

        var activeKcal: Int?
        var totalKcal: Int?
        if let dual = text.range(of: #"Actieve\s*kilocalorie[ëe]n\s*Totale\s*kilocalorie[ëe]n\s*(\d+)\s*KCAL\s*(\d+)\s*KCAL"#, options: [.regularExpression, .caseInsensitive]) {
            let token = String(text[dual])
            let numbers = token.components(separatedBy: CharacterSet.decimalDigits.inverted).compactMap { Int($0) }
            if numbers.count >= 2 {
                activeKcal = numbers[numbers.count - 2]
                totalKcal = numbers[numbers.count - 1]
            }
        } else {
            activeKcal = firstMatch(text, patterns: [#"Actieve\s*kilocalorie[ëe]n\s*(\d+)"#, #"Active\s*calories\s*(\d+)"#]).flatMap { Int($0) }
            totalKcal = firstMatch(text, patterns: [#"Totale\s*kilocalorie[ëe]n\s*(\d+)"#, #"Total\s*calories\s*(\d+)"#]).flatMap { Int($0) }
        }
        if activeKcal != nil { fieldsFound += 1 }
        if totalKcal != nil { fieldsFound += 1 }

        let paceSecPer100m = firstMatch(text, patterns: [
            #"Gem\.?\s*tempo\s*Gem\.?\s*hartslag\s*(\d+'\d{1,2}")"#,
            #"Gem\.?\s*tempo\s*(\d+'\d{1,2}")"#,
            #"Gem\.?\s*tempo\s*Gem\.?\s*hartslag\s*(\d{2,3}")"#,
            #"(\d+'\d{1,2}")(?:\s*/\s*100\s*m|\s*\d*\s*m)"#,
            #"(\d{2,3}")(?:\s*[n\/]?\d*\s*m|\s*SPM)"#
        ]).flatMap { parseOcrPace($0) }
        if paceSecPer100m != nil { fieldsFound += 1 }

        let avgHeartRate = firstMatch(text, patterns: [#"(\d{2,3})\s*SPM"#, #"Gem\.?\s*hartslag\s*(\d{2,3})"#]).flatMap { Int($0) }
        if avgHeartRate != nil { fieldsFound += 1 }

        var laps: Int?
        var poolLengthM = 25
        if let lapsPool = text.range(of: #"Banen\s*Baanlengte\s*(\d+)\s*(\d+)\s*M"#, options: [.regularExpression, .caseInsensitive]) {
            let token = String(text[lapsPool])
            let numbers = token.components(separatedBy: CharacterSet.decimalDigits.inverted).compactMap { Int($0) }
            if numbers.count >= 2 {
                laps = numbers[numbers.count - 2]
                poolLengthM = numbers[numbers.count - 1]
            }
        } else {
            laps = firstMatch(text, patterns: [#"Banen\s*(\d+)"#, #"Laps\s*(\d+)"#]).flatMap { Int($0) }
            if let pool = firstMatch(text, patterns: [#"Baanlengte\s*(\d+)\s*M"#, #"\b(\d{2})\s*M\s*$"#]).flatMap({ Int($0) }) {
                poolLengthM = pool
            }
        }
        if laps != nil { fieldsFound += 1 }

        let goalM = firstMatch(text, patterns: [#"Doel:\s*([\d.]+)\s*M"#]).flatMap { SwimFormatters.parseDistanceM($0) }
        if goalM != nil { fieldsFound += 1 }

        let timeRange: String
        if let range = text.range(of: #"\b(\d{1,2}:\d{2})\s*-\s*(\d{1,2}:\d{2})\b"#, options: .regularExpression) {
            let token = String(text[range])
            let parts = token.split(separator: "-").map { String($0).trimmingCharacters(in: .whitespaces) }
            timeRange = parts.count == 2 ? "\(parts[0])–\(parts[1])" : ""
        } else {
            timeRange = ""
        }

        var location = ""
        if text.localizedCaseInsensitiveContains("Tilburg") {
            location = "Tilburg"
        }

        let strokes = StrokeDistances(
            mixedM: parseStrokeDistance(text, label: "Gemengd"),
            breaststrokeM: parseStrokeDistance(text, label: "Schoolslag"),
            freestyleM: parseStrokeDistance(text, label: "Vrije\\s*slag"),
            backstrokeM: parseStrokeDistance(text, label: "Rugslag"),
            butterflyM: parseStrokeDistance(text, label: "Vlinderslag")
        )
        if [strokes.mixedM, strokes.breaststrokeM, strokes.freestyleM, strokes.backstrokeM, strokes.butterflyM].contains(where: { $0 != nil }) {
            fieldsFound += 1
        }

        if distanceM == nil { warnings.append("distance_not_found") }
        if paceSecPer100m == nil { warnings.append("pace_not_found") }
        if missingDate { warnings.append("date_missing") }

        let fields = ParsedScreenshotFields(
            date: date,
            durationSec: durationSec,
            distanceM: distanceM,
            activeKcal: activeKcal,
            totalKcal: totalKcal,
            paceSecPer100m: paceSecPer100m,
            avgHeartRate: avgHeartRate,
            laps: laps,
            poolLengthM: poolLengthM,
            goalM: goalM,
            location: location,
            timeRange: timeRange,
            strokes: strokes
        )

        let detection = detectSwimWorkout(text: text, fields: fields)
        if !detection.isSwimWorkout {
            warnings.append("not_swim_workout")
        }

        let confidence = min(100, Int(round(Double(fieldsFound) / 10.0 * 100.0)))

        return ParsedScreenshotResult(
            fields: fields,
            confidence: confidence,
            missingDate: missingDate,
            warnings: warnings,
            isSwimWorkout: detection.isSwimWorkout,
            detectedSport: detection.detectedSport
        )
    }

    static func fieldsToMetrics(_ fields: ParsedScreenshotFields) -> SwimMetrics {
        SwimMetrics(
            durationSec: fields.durationSec,
            distanceM: fields.distanceM,
            activeKcal: fields.activeKcal,
            totalKcal: fields.totalKcal,
            paceSecPer100m: fields.paceSecPer100m,
            avgHeartRate: fields.avgHeartRate,
            laps: fields.laps,
            poolLengthM: fields.poolLengthM,
            goalM: fields.goalM,
            location: fields.location,
            timeRange: fields.timeRange,
            strokes: fields.strokes
        )
    }

    private static func firstMatch(_ text: String, patterns: [String]) -> String? {
        for pattern in patterns {
            guard let range = text.range(of: pattern, options: [.regularExpression, .caseInsensitive]) else { continue }
            let match = String(text[range])
            if let capture = captureGroup(from: match, pattern: pattern) {
                return capture
            }
        }
        return nil
    }

    private static func captureGroup(from match: String, pattern: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return nil }
        let nsRange = NSRange(match.startIndex..<match.endIndex, in: match)
        guard let result = regex.firstMatch(in: match, range: nsRange), result.numberOfRanges > 1,
              let range = Range(result.range(at: 1), in: match) else {
            return nil
        }
        return String(match[range])
    }

    private static func parseDistanceToken(_ token: String) -> Int? {
        SwimFormatters.parseDistanceM(token.replacingOccurrences(of: " ", with: ""))
    }

    private static func parseOcrPace(_ raw: String) -> Int? {
        if let direct = SwimFormatters.parsePaceSecPer100m(raw) { return direct }
        if let match = raw.range(of: #"^(\d{2,3})""#, options: .regularExpression) {
            let digits = String(raw[match]).filter(\.isNumber)
            if digits.count == 3 {
                let minutes = Int(String(digits.prefix(1))) ?? 0
                let seconds = Int(String(digits.suffix(2))) ?? 0
                return minutes * 60 + seconds
            }
        }
        return nil
    }

    private static func parseStrokeDistance(_ text: String, label: String) -> Int? {
        let pattern = "\(label)\\s*\\(([\\d.\\s]+)\\s*m\\)"
        guard let match = firstMatch(text, patterns: [pattern]) else { return nil }
        return SwimFormatters.parseDistanceM(match)
    }

    private static func detectSwimWorkout(text: String, fields: ParsedScreenshotFields) -> (isSwimWorkout: Bool, detectedSport: String?) {
        let otherSports: [(String, [String])] = [
            ("run", [#"\bHardlopen\b"#, #"\bIndoor\s*Run\b"#]),
            ("walk", [#"\bWandelen\b"#, #"\bOutdoor\s*Walk\b"#]),
            ("cycle", [#"\bFietsen\b"#, #"\bIndoor\s*Cycling\b"#]),
            ("strength", [#"\bKrachttraining\b"#])
        ]

        for (sport, patterns) in otherSports {
            if patterns.contains(where: { text.range(of: $0, options: [.regularExpression, .caseInsensitive]) != nil }) {
                return (false, sport)
            }
        }

        var swimSignals = 0
        if text.range(of: #"\bZwem(bad|men)\b"#, options: [.regularExpression, .caseInsensitive]) != nil { swimSignals += 1 }
        if text.range(of: #"\bBanen\b"#, options: .regularExpression) != nil { swimSignals += 1 }
        if text.range(of: #"/\s*100\s*m\b"#, options: .regularExpression) != nil { swimSignals += 1 }
        if fields.paceSecPer100m != nil { swimSignals += 1 }
        if fields.laps != nil { swimSignals += 1 }
        if [fields.strokes.mixedM, fields.strokes.breaststrokeM, fields.strokes.freestyleM, fields.strokes.backstrokeM, fields.strokes.butterflyM].contains(where: { $0 != nil }) {
            swimSignals += 1
        }

        if swimSignals >= 2 {
            return (true, nil)
        }
        return (false, "unknown")
    }
}
